FP16 inference performance improvement on  CPU

nikhilfujitsu opened the pull request:
Author: Masaru Ito [masaru.ito.mi@jp.fujitsu.com](mailto:masaru.ito.mi@jp.fujitsu.com)

### Description
1.Added fp16support for add,sub,mul and div operators thus enable gelu fusion and erf and surrounding ops
2.Enable fp16 to fp32 cast(performance improved from 15sec -> 0.27sec)
3.Added eigen::fp16 support in layer normalization (performance improved from 22sec to 0.6sec)
4.Enable fp16 Transpose call in mlas(performance improved from 3sec -> 0.47sec​)

Rohanjames1997 commented:
Hi @nikhilfujitsu, these are great perf gains!

Could you clarify on which CPU(s) did you witness these perf improvements?
And do you have steps to reproduce them?

abhijain1204fujitsu commented:
> Hi @nikhilfujitsu, these are great perf gains!
> 
> Could you clarify on which CPU(s) did you witness these perf improvements? And do you have steps to reproduce them?

Hi @Rohanjames1997 the results were captured using AWS Graviton 3e Machine (c7gn), you can try to run on the same

Rohanjames1997 commented:
Thanks @abhijain1204fujitsu , I will try it out later. 
If you could share the steps to repro the speedup - like `onnxruntime_perf_test` or the like, that would help too.

nikhilfujitsu commented:
@edgchen1 sir, please support this PR.

edgchen1 commented on onnxruntime/core/providers/cpu/math/element_wise_ops.cc:
```diff
@@ -1923,10 +1992,28 @@ Status Erf<float>::Compute(OpKernelContext* context) const {
 
+template <>
+Status Erf<MLFloat16>::Compute(OpKernelContext* context) const {
   return Status::OK();
```
why does `Erf<MLFloat16>::Compute()` do nothing?

edgchen1 commented on onnxruntime/core/providers/cpu/nn/layer_norm_impl.cc:
```diff
@@ -11,6 +11,7 @@
 #include "core/providers/common.h"
 #include "core/util/force_inline.h"
 #include "core/util/math_cpuonly.h"
+#include <Eigen/Core>
```
`<Eigen/Core>` is already included in `core/util/math_cpuonly.h`. there, it also has some handling for warnings. I'd suggest just using that so we don't need to duplicate the warning suppressions here too.

edgchen1 commented on onnxruntime/core/providers/cpu/nn/layer_norm_impl.cc:
```diff
@@ -80,6 +81,11 @@ void ComputeJob(
   }
 }
 
+// Helper to convert int64_t -> Eigen::Index safely
+inline Eigen::Index ToEigenIndex(int64_t v) {
+  return static_cast<Eigen::Index>(v);
+}
```
I wouldn't consider `static_cast` a safe conversion. a checked conversion like `narrow` would be safer.

nikhilfujitsu replied:
Will incorporate.

edgchen1 commented on onnxruntime/core/providers/cpu/tensor/cast_op.cc:
```diff
@@ -390,11 +391,32 @@ struct TensorCaster<MLFloat16, float> {
     const size_t shape_size = narrow<size_t>(shape.Size());
-    MlasConvertFloatToHalfBuffer(in_data, out_data, shape_size);
+
+    // Check if the tensor is long enough to use threads
+    if (shape_size <= 128000) {
+      MlasConvertFloatToHalfBuffer(in_data, out_data, shape_size);
+      return;
+    }
+    // Calculate the number of compute cyles per implementation
+    auto cpu_info = CPUIDInfo::GetCPUIDInfo();
+    double num_compute_cycles;
+    if (cpu_info.HasSSE3()) {
+      num_compute_cycles = static_cast<double>(shape_size >> 1);
+    } else if (cpu_info.HasAVX2()) {
+      num_compute_cycles = static_cast<double>(shape_size >> 2);
+    } else {
+      num_compute_cycles = static_cast<double>(shape_size * 10);
+    }
```
I think these platform-specific details should be within the MLAS code instead of an individual op implementation. in the other direction, we have `MlasConvertHalfToFloatBufferInParallel()`. maybe we should create an analogous `MlasConvertFloatToHalfBufferInParallel()`.

https://github.com/microsoft/onnxruntime/blob/af4bf436d70f62afe486bb5b9d9cc7c8f3b5b958/onnxruntime/core/mlas/lib/cast.cpp#L37-L88

edgchen1 commented on onnxruntime/test/providers/cpu/math/element_wise_ops_test.cc:
```diff
@@ -3941,6 +3942,30 @@ TEST(MathOpTest, ErfMoreData) {
+  std::vector<float> outputs(size);
+
+  float low = -5.0f, high = 5.0f;
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_real_distribution<float> dist(low, high);
```
consider using the `RandomValueGenerator` class. it shows the random seed on failure so failures are easier to reproduce.

https://github.com/microsoft/onnxruntime/blob/af4bf436d70f62afe486bb5b9d9cc7c8f3b5b958/onnxruntime/test/common/random_generator.h#L43

also consider using a fixed random seed to avoid flakiness in CI builds.

nikhilfujitsu commented on docs/OperatorKernels.md:
```diff
@@ -5,8 +5,6 @@ Do not modify directly.*
 ## Execution Providers
 
 - [CPUExecutionProvider](#cpuexecutionprovider)
-- [CUDAExecutionProvider](#cudaexecutionprovider)
```
Will fix this.

nikhilfujitsu commented:
@edgchen1 sir please support this PR further. Thank you for giving your valuable time.

edgchen1 commented on onnxruntime/core/mlas/lib/cast.cpp:
```diff
@@ -71,16 +71,16 @@ MlasConvertHalfToFloatBufferInParallel(
-                static_cast<double>(Count * sizeof(MLAS_FP16)),  // Size of no. of elements in bytes to be loaded
-                static_cast<double>(Count * sizeof(float)),      // Size of no. of elements in bytes to be stored
-                static_cast<double>(num_compute_cycles),         // No. of compute cycles required for the tensor op
+                gsl::narrow<double>(Count * sizeof(MLAS_FP16)),  // Size of no. of elements in bytes to be loaded
```
generally, we should use `onnxruntime::narrow` from include/onnxruntime/core/common/narrow.h. it also supports builds where exceptions are disabled.

edgchen1 commented on onnxruntime/core/providers/cpu/cpu_execution_provider.cc:
```diff
@@ -1426,2032 +1440,2037 @@ KernelCreateInfo BuildKernelCreateInfo<void>() {
 
 Status RegisterOnnxOperatorKernels(KernelRegistry& kernel_registry) {
   static const BuildKernelCreateInfoFn function_table[] = {
-      BuildKernelCreateInfo<void>,  // default entry to avoid the list become empty after ops-reducing
```
there are a lot of whitespace changes in this file. please ensure that it is formatted (the CI build will also complain about this).

you can use lintrunner. more details here:
https://github.com/microsoft/onnxruntime/blob/main/docs/Coding_Conventions_and_Standards.md#linting

edgchen1 commented on onnxruntime/core/providers/cpu/math/element_wise_ops.cc:
```diff
@@ -590,6 +595,23 @@ Status Add<T>::Compute(OpKernelContext* context) const {
 
+template <>
+Status Add<MLFloat16>::Compute(OpKernelContext* context) const {
+  // BroadcastHelper received as argument may differ from 'helper' when parallelizing within a span
```
what is the `helper` referred to by this comment?

nikhilfujitsu replied:
The comment means “don’t assume per_iter_bh is the same as the original broadcast helper (helper). When parallel execution is used, this helper may be split/adjusted to point only to the span of data handled by this thread.”

edgchen1 replied:
oh. it looks like this comment was copied from the `Add<T>` template above. possibly there used to be a variable called `helper` in scope but the comment wasn't updated.

nikhilfujitsu replied:
sure will remove it.

edgchen1 commented on onnxruntime/test/optimizer/graph_transform_test.cc:
```diff
@@ -672,7 +672,7 @@ TEST_F(GraphTransformationTests, ConstantFoldingUnsupportedFloat16) {
 
   op_to_count = CountOpsInGraph(graph);
-  ASSERT_TRUE(op_to_count["Mul"] == 1);
+  ASSERT_TRUE(op_to_count["Mul"] == 0);  // Mul is folded
```
from the name `ConstantFoldingUnsupportedFloat16`, it seems that this test is supposed to test a case where constant folding doesn't happen. perhaps the test can be updated so that it still covers an unsupported case, or if it no longer applies, it can be removed.

nikhilfujitsu replied:
Sure will remove it.

edgchen1 replied:
sorry, what I meant was removing the `TEST_F(GraphTransformationTests, ConstantFoldingUnsupportedFloat16)` test entirely if it is no longer applicable.

edgchen1 commented on onnxruntime/core/providers/cpu/math/element_wise_ops.cc:
```diff
@@ -1923,9 +1992,57 @@ Status Erf<float>::Compute(OpKernelContext* context) const {
+        // Compute ERF in float32
+        MlasComputeErf(input_fp32.data(), output_fp32.data(), narrow<std::ptrdiff_t>(count));
+        // Convert output from float32 to MLFloat16 (bulk, SIMD-capable)
+        MlasConvertFloatToHalfBuffer(output_fp32.data(), p_output, narrow<std::ptrdiff_t>(count));
```
nit: `narrow` is not free - it might be good to save and reuse the result or just work with the correct type from the beginning.

nikhilfujitsu replied:
Sure will incorporate this.

nikhilfujitsu commented on cmake/external/onnx:
Yes.

nikhilfujitsu commented:
@edgchen1 sir please support this PR further. Thanks.

hariharans29 commented:
> @edgchen1 sir please support this PR further. Thanks.

The ONNX submodule update doesn't seem to be removed.

nikhilfujitsu commented:
> > @edgchen1 sir please support this PR further. Thanks.
> 
> The ONNX submodule update doesn't seem to be removed.

Updated the cmake/externel/onnx.

edgchen1 commented on onnxruntime/test/optimizer/graph_transform_test.cc:
```diff
@@ -649,37 +649,6 @@ TEST_F(GraphTransformationTests, ConstantFoldingTransposeEmptyInitializer) {
 }
 
-TEST_F(GraphTransformationTests, ConstantFoldingUnsupportedFloat16) {
-  constexpr const ORTCHAR_T* model_uri = MODEL_FOLDER "constant_float16_mul.onnx";
```
I think this model file was only used by this test. if you are deleting it, please also clean up this file.

looking at it more though, it looks like this unit test was added as part of a fix in this PR.
https://github.com/microsoft/onnxruntime/pull/17399

I wonder if we should instead adapt it so that it will still work as a regression test. perhaps find another op/type that is not supported by the CPU EP and use that instead?

nikhilfujitsu replied:
Okay currently the following operators shows that they don't have the support of MLfloat16 
Tanh
TfIdfVectorizer
ThresholdedRelu
Tile
TopK
Transpose
TreeEnsemble
TreeEnsembleClassifier
TreeEnsembleRegressor
Trilu
Unique
Unsqueeze
Where
Xor
ZipMap

I can change this regression testing for tanh.

edgchen1 commented on onnxruntime/test/optimizer/graph_transform_test.cc:
```diff
@@ -649,32 +649,32 @@ TEST_F(GraphTransformationTests, ConstantFoldingTransposeEmptyInitializer) {
 
-  // assign all nodes to CUDA. the constant folding should try folding the node on the CPU and fail, thus leaving the
-  // EP as CUDA and not constant folding the node.
+  // assign all nodes to CUDA. constant folding will attempt CPU but fail, leaving nodes on CUDA.
```
please consider this update. looks like the comment just got resolved without a reply or change.

edgchen1 commented:
/azp run Linux QNN CI Pipeline,Win_TRT_Minimal_CUDA_Test_CI,Windows ARM64 QNN CI Pipeline,Windows GPU Doc Gen CI Pipeline

hariharans29 commented:
Please run onnxruntime_mlas_test in your env with your change and report any failures if any here.

edgchen1 commented on onnxruntime/core/providers/cpu/nn/layer_norm_impl.cc:
```diff
@@ -96,58 +101,61 @@ void ComputeJob(
-    mean += input_float_ptr[h];
-    mean_square += input_float_ptr[h] * input_float_ptr[h];
+  for (int64_t i = 0; i < norm_size; ++i) {
+    float val = gsl::narrow_cast<float>(input_vec[ToEigenIndex(i)]);
```
general nit: it's not a narrowing conversion from `Eigen::half` to `float`. maybe `static_cast` would be clearer.

nikhilfujitsu commented:
> > Build and install the wheel of Onnxruntime. Download the INT/Float E5 model from Hugging Face. Convert the model to FP16 ONNX. Create a Python Script to Run Inference. To Analyze Operator-Level Timings use Session option.
> > We used AWS Graviton 3e Machine (c7gn) with 64 cores.
> 
> thanks. would you mind also updating the PR description to include this information, so it is easier to find later?
> 
> also, regarding these measurements:
> 
> > 2.Enable fp16 to fp32 cast(performance improved from 15sec -> 0.27sec)
> > 3.Added eigen::fp16 support in layer normalization (performance improved from 22sec to 0.6sec)
> > 4.Enable fp16 Transpose call in mlas(performance improved from 3sec -> 0.47sec​)
> 
> could you please clarify what is being measured? are the timings for different sections of the FP16 E5 model?

I have updated the Description.
During the inference operation only the time is captured. Yes, the timings are for the different ops for the E5 model.

nikhilfujitsu commented:
Hi @hariharans29 sir,
Seems like build is okay but issues with CI.
The repository's CI requirements file pins onnxscript==0.3.2 (tools/ci_build/github/linux/python/requirements.txt in ref 4f238652eb39bc8381fa41b4744aa647961ad285), which appears incompatible with the runner's torch version.

nikhilfujitsu commented:
@hariharans29  sir please restart the CI.

nikhilfujitsu commented on onnxruntime/core/providers/cpu/nn/layer_norm_impl.cc:
```diff
@@ -96,58 +101,61 @@ void ComputeJob(
-  float mean(0.0f);
-  float mean_square(0.0f);
+  // reinterpret input/output MLFloat16* as Eigen::half*
+  const Eigen::half* p_input = reinterpret_cast<const Eigen::half*>(X_data + task_idx * norm_size);
```
No behavioural changes apart from minor rounding differences.

nikhilfujitsu replied:
I am not able to get what exactly you are looking in. I have answered this. Please help me here.

nikhilfujitsu commented on onnxruntime/core/providers/cpu/math/element_wise_ops.cc:
```diff
@@ -1923,9 +1991,64 @@ Status Erf<float>::Compute(OpKernelContext* context) const {
+        const float* p_input = input_data + start;
+        float* p_output = output_data + start;
+        int64_t count = std::min(length_per_task, elem_count - start);
+        MlasComputeErf(p_input, p_output, narrow<std::ptrdiff_t>(count));
```
ok

nikhilfujitsu commented:
@edgchen1 Please support this PR further.
Thanks.

hariharans29 commented:
There are some open comments left to be addressed. Please address them.

https://github.com/microsoft/onnxruntime/pull/25680#discussion_r2415316277
https://github.com/microsoft/onnxruntime/pull/25680#discussion_r2441385843
https://github.com/microsoft/onnxruntime/pull/25680#discussion_r2443472735
https://github.com/microsoft/onnxruntime/pull/25680#discussion_r2443473256
https://github.com/microsoft/onnxruntime/pull/25680#discussion_r2443473322

hariharans29 commented on onnxruntime/core/mlas/lib/cast.cpp:
```diff
@@ -80,7 +81,7 @@ MlasConvertHalfToFloatBufferInParallel(
                     Source + first_span,
                     Destination + first_span,
-                    static_cast<size_t>(last_span - first_span));
+                    onnxruntime::narrow<std::ptrdiff_t>(last_span - first_span));
```
Revert and mirror the same in the new routine

nikhilfujitsu replied:
Done.

edgchen1 commented on onnxruntime/core/mlas/lib/cast.cpp:
```diff
@@ -14,6 +14,7 @@ Module Name:
 
 --*/
 #include "mlasi.h"
+#include "core/common/narrow.h"
```
nit: looks like we don't need this include now
